lib/proxy/peer/internal/tls.go

```diff
@@ -0,0 +1,97 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package internal
+
+import (
+	"context"
+	"crypto/x509"
+	"log/slog"
+	"slices"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// VerifyPeerCertificateIsProxy is a function usable as a
+// [tls.Config.VerifyPeerCertificate] callback to enforce that the connected TLS
+// client is using Proxy credentials.
+func VerifyPeerCertificateIsProxy(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	if len(verifiedChains) < 1 {
+		return trace.AccessDenied("missing client certificate (this is a bug)")
+	}
+
+	clientCert := verifiedChains[0][0]
```

Comment:
Should we check `len(verifiedChains[0])` first, just to be safe?

Reply:
`verifiedChains` is a slice of nonempty slices, as per the `crypto/tls` docs. I think I'd rather segfault than live in a world where my TLS library is violating an invariant around certificate chain verification - wdyt?